### PR TITLE
Add Ml shared preferences for custom model storage

### PR DESCRIPTION
**Pull request by annzimmer**

### CODE REVIEW

File: firebase-ml-modeldownloader/src/test/java/com/google/firebase/ml/modeldownloader/internal/SharedPreferencesUtilTest.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+  private static final CustomModel CUSTOM_MODEL_DOWNLOAD_COMPLETE =
+      new CustomModel(MODEL_NAME, 0, 100, MODEL_HASH, "file/path/store/ModelName/1");
+
+  private static final CustomModel CUSTOM_MODEL_UPDATE_IN_BACKGROUND =
+      new CustomModel(MODEL_NAME, 986, 100, MODEL_HASH, "file/path/store/ModelName/1");
+
+  private static final CustomModel CUSTOM_MODEL_DOWNLOADING =
+      new CustomModel(MODEL_NAME, 986, 100, MODEL_HASH);
+
+  //  private static final FirebaseApp FIREBASE_APP = FirebaseApp.initializeApp(
```
Comment (rlazo) [inline firebase-ml-modeldownloader/src/test/java/com/google/firebase/ml/modeldownloader/internal/SharedPreferencesUtilTest.java:47]: Can this code be removed?

File: firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/internal/SharedPreferencesUtil.java [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+    String filePath = customModel.getLocalFilePath();
+    getSharedPreferences()
+        .edit()
+        .putString(String.format(LOCAL_MODEL_HASH_PATTERN, persistenceKey, modelName), hash)
+        .putLong(String.format(LOCAL_MODEL_FILE_SIZE_PATTERN, persistenceKey, modelName), size)
+        .putString(
+            String.format(LOCAL_MODEL_FILE_PATH_PATTERN, persistenceKey, modelName), filePath)
+        .apply();
+
+    clearDownloadingModelDetails(customModel.getName());
```
Comment (rlazo) [inline firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/internal/SharedPreferencesUtil.java:170]: Would it be useful to do the clearing as part of the change above? If yes, then you can pass the editor to `clearDownloadingModelDetails` and use it to add the remove statements in a single `apply()` call.

File: firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/internal/SharedPreferencesUtil.java [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+            String.format(DOWNLOADING_MODEL_HASH_PATTERN, persistenceKey, modelName), modelHash)
+        .putLong(
+            String.format(DOWNLOADING_MODEL_SIZE_PATTERN, persistenceKey, modelName), modelSize)
+        .putLong(String.format(DOWNLOADING_MODEL_ID_PATTERN, persistenceKey, modelName), downloadId)
+        // The following assumes the download will finish before the system reboots.
+        // If not, the download duration won't be correct, which isn't critical.
+        .putLong(
+            String.format(DOWNLOAD_BEGIN_TIME_MS_PATTERN, persistenceKey, modelName),
+            SystemClock.elapsedRealtime())
+        .apply();
```
Comment (rlazo) [inline firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/internal/SharedPreferencesUtil.java:141]: `apply()` only alters the in-memory contents of the SharedPreferences, and eventually writes to disk, which may fail silently, unlike `commit()` that's synchronous. Just want to double check that this is an acceptable situation in this case.

Comment (annzimmer) [reply firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/internal/SharedPreferencesUtil.java]: changed to commit

File: firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/internal/SharedPreferencesUtil.java (hunk 2 above)
Comment (annzimmer) [reply firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/internal/SharedPreferencesUtil.java]: great idea.

File: firebase-ml-modeldownloader/src/test/java/com/google/firebase/ml/modeldownloader/internal/SharedPreferencesUtilTest.java (hunk 1 above)
Comment (annzimmer) [reply firebase-ml-modeldownloader/src/test/java/com/google/firebase/ml/modeldownloader/internal/SharedPreferencesUtilTest.java]: done.

File: firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/internal/SharedPreferencesUtil.java [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    this.firebaseApp = firebaseApp;
+    this.persistenceKey = firebaseApp.getPersistenceKey();
+  }
+
+  /**
+   * Returns the Custom Model details currently associated with this model. If a fully downloaded
+   * model is present - this returns the details of that model, including local file path. If an
+   * update of an existing model is in progress, the local model plus the download id for the new
+   * upload is returned. To get only details related to the downloading model use
+   * getDownloadingCustomModelDetails. If this is the initial download of a local file - the
```
[the remark this answers is not in the document]
Comment (annzimmer) [reply firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/internal/SharedPreferencesUtil.java]: done

File: firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/CustomModel.java [hunk 5]
```diff
@@ -76,18 +118,31 @@ public long getDownloadId() {
     return downloadId;
   }
 
+  @Override
+  public boolean equals(Object o) {
```
[the remark this answers is not in the document]
Comment (annzimmer) [reply firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/CustomModel.java:118]: thanks - added.

File: firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/CustomModel.java [hunk 6]
```diff
@@ -28,7 +29,48 @@
...
+   *
+   * @param name - model name
+   * @param downloadId - Android Download Manger - download id
+   * @param fileSize - model file size
+   * @param modelHash - model hash size
+   * @hide
+   */
+  public CustomModel(
+      @NonNull String name, long downloadId, long fileSize, @NonNull String modelHash) {
+    this.modelHash = modelHash;
```
[the remark this answers is not in the document]
Comment (annzimmer) [reply firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/CustomModel.java]: done.

**Review by rlazo — APPROVED** [review]

**Pull request merged**